user/test/com/google/gwt/emultest/java11/lang/StringTest.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.gwt.emultest.java11.lang;
+
+import com.google.gwt.emultest.java.util.EmulTestBase;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+/**
+ * Tests for java.lang.String Java 11 API emulation.
+ */
+public class StringTest extends EmulTestBase {
+    public void testIsBlank() {
+        assertTrue("".isBlank());
+        assertTrue("  ".isBlank());
+        assertFalse("x ".isBlank());
+        assertTrue("\u001c".isBlank());
+        assertFalse("\u00a0".isBlank());
+    }
+
+    public void testStrip() {
```

Comment:
Esp given https://github.com/gwtproject/gwt/issues/9973, would you mind also checking that the trim() tests correctly leave these non-ascii whitespace chars intact?

Also, from the same issue, should we have tests for `\ufeff` and `\u180e` to ensure they are preserved? Granted they would fail at this time, but might be good to have them while touching relevant tests (and comment them out, link them to the new issue).

Reply:
I added more testcases, then I stumbled upon
```

  /**
   * Helper method for testTrim to avoid compiler optimizations.
   *
   * TODO: insufficient, compiler now inlines.
   */
  public void trimRightAssertEquals(String left, String right) {
    assertEquals(left, right.trim());
  }
  ```
  Is there a way to check the compiler output to see if the test is actually testing the emulation?

Reply:
Good catch, here's what I tend to do: 
 * add `-setProperty compiler.useSourceMaps=true` to `test.args`, this disables the stack trace emulation that makes the output hard to read. Note that this may slightly break your stack traces, the line numbers won't quite be accurate (but close)
 * add `-style PRETTY` also to `test.args`, this makes the output more or less readable
 * add some apparent nondeterminism, or go through JS to your code to make the compiler unable to optimize - for example, an always-true condition like `Math.random() < 5` in an `if` or ternary, or a JSNI unchecked cast wrapping a value

From the first two, when I look at the JS file in build/out/user/test/web-htmlunit/www/com.google.gwt.emultest.EmulSuite.JUnit/ I see
```
_.testStrip = function testStrip(){
  assertEquals_1('', '');
  assertEquals_1('', '');
  assertEquals_1('x', 'x');
  assertEquals_1('x', 'x');
  assertEquals_1('\xA0x\xA0', '\xA0x\xA0');
  assertEquals_1('\uFFEFx\u180E', '\uFFEFx\u180E');
}
```
Clearly that's no good - testRepeat at least is the same:
```
_.testRepeat = function testRepeat(){
  var ex;
  assertEquals_1('', '');
  assertEquals_1('foo', 'foo');
  assertEquals_1('foofoofoo', 'foofoofoo');
  try {
    checkCriticalArgument(false, 'count is negative: -1');
    'foo'.repeat(-1);
    throw toJs(new Error_0('Should fail with negative arg'));
  }
  // ...
```

Note however that the draft tests do better, since the compiler doesnt optimize as heavily:
```
_.stripRightAsssertEquals = function stripRightAsssertEquals(expected, arg){
  jf.assertEquals_1(expected, jl.strip__Ljava_lang_String___devirtual$(arg));
}
;
//...
_.testStrip = function testStrip(){
  this.stripRightAsssertEquals('', '');
  this.stripRightAsssertEquals('', '  ');
  this.stripRightAsssertEquals('x', ' x ');
  this.stripRightAsssertEquals('x', '\x1Cx\x1C');
  this.stripRightAsssertEquals('\xA0x\xA0', '\xA0x\xA0 ');
  this.stripRightAsssertEquals('\uFFEFx\u180E', '\uFFEFx\u180E ');
}
;
```
```
_.testRepeat = function testRepeat(){
  var ex, noFoo;
  jf.assertEquals_1('', jl.repeat_I_Ljava_lang_String___devirtual$('foo', 0));
  jf.assertEquals_1('foo', jl.repeat_I_Ljava_lang_String___devirtual$('foo', 1));
  jf.assertEquals_1('foofoofoo', jl.repeat_I_Ljava_lang_String___devirtual$('foo', 3));
```

This is caused by DeadCodeElimination.tryOptimizeStringCall, where it finds a string literal with a plain method call on it, and calls it reflectively in the JVM at compile time, saving the compiled JS from needing to do it.

So we need to "hide" the string constants from the compiler so it doesnt just call trim() or repeat() on it, for example, StringTest.java does this (but doesn't use it for trim tests?):
https://github.com/gwtproject/gwt/blob/4aee3bd9df97e2b96a444751ea26a3a9091223f5/user/test/com/google/gwt/emultest/java/lang/StringTest.java#L957-L963

Used like this:
```
  private void stripRightAsssertEquals(String expected, String arg) {
    assertEquals(expected, hideFromCompiler(arg).strip());
  }
```
This makes for some awful output:
```
_.testStrip = function testStrip(){
  assertEquals_1('', $strip(($wnd.Math.random() < -1 && fail() , '')));
  assertEquals_1('', $strip(($wnd.Math.random() < -1 && fail() , '  ')));
  assertEquals_1('x', $strip(($wnd.Math.random() < -1 && fail() , ' x ')));
  assertEquals_1('x', $strip(($wnd.Math.random() < -1 && fail() , '\x1Cx\x1C')));
  assertEquals_1('\xA0x\xA0', $strip(($wnd.Math.random() < -1 && fail() , '\xA0x\xA0 ')));
  assertEquals_1('\uFFEFx\u180E', $strip(($wnd.Math.random() < -1 && fail() , '\uFFEFx\u180E ')));
}
;
_.testStripLeading = function testStripLeading(){
  assertEquals_1('', $stripLeading(($wnd.Math.random() < -1 && fail() , '')));
  assertEquals_1('', $stripLeading(($wnd.Math.random() < -1 && fail() , '  ')));
  assertEquals_1('x ', $stripLeading(($wnd.Math.random() < -1 && fail() , ' x ')));
  assertEquals_1('x\x1C', $stripLeading(($wnd.Math.random() < -1 && fail() , '\x1Cx\x1C')));
  assertEquals_1('\xA0x\xA0', $stripLeading(($wnd.Math.random() < -1 && fail() , '\xA0x\xA0')));
}
;
_.testStripTrailing = function testStripTrailing(){
  assertEquals_1('', $stripTrailing(($wnd.Math.random() < -1 && fail() , '')));
  assertEquals_1('', $stripTrailing(($wnd.Math.random() < -1 && fail() , '  ')));
  assertEquals_1(' x', $stripTrailing(($wnd.Math.random() < -1 && fail() , ' x ')));
  assertEquals_1('\x1Cx', $stripTrailing(($wnd.Math.random() < -1 && fail() , '\x1Cx\x1C')));
  assertEquals_1('\xA0x\xA0', $stripTrailing(($wnd.Math.random() < -1 && fail() , '\xA0x\xA0 ')));
}
;
```
but we can clearly see that it is no longer able to optimize away the call. I'll put up the change for the plain StringTest class, no idea how that was ignored for so long...